src/main/java/com/linkedin/kmf/services/KeyedConsumeService.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.linkedin.kmf.services;
+
+import com.linkedin.kmf.common.DefaultTopicSchema;
+import com.linkedin.kmf.common.Utils;
+import com.linkedin.kmf.consumer.BaseConsumerRecord;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.avro.generic.GenericRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class KeyedConsumeService extends ConsumeService implements Service {
+  private static final Logger LOG = LoggerFactory.getLogger(KeyedConsumeService.class);
+
+  public KeyedConsumeService(Map<String, Object> props, String name) throws Exception {
+    super(props, name);
+  }
+
+  private void consume() throws Exception {
+    Map<String, Long> nextIndexes = new HashMap<>();
+
+    while (_running.get()) {
+      BaseConsumerRecord record;
+      try {
+        record = _consumer.receive();
+      } catch (Exception e) {
+        _sensors._consumeError.record();
+        LOG.warn(_name + "/KeyedConsumeService failed to receive record", e);
+        // Avoid busy while loop
+        Thread.sleep(100);
+        continue;
+      }
+
+      if (record == null)
+        continue;
+
+      GenericRecord avroRecord = Utils.genericRecordFromJson(record.value());
+      if (avroRecord == null) {
+        _sensors._consumeError.record();
+        continue;
+      }
+      String key = record.key();
```

Comment:
Have a protected method like Object sequenceNumberKeyFromRecord(GenericRecord consumed) that returns the key used in the hash map.   This can be string or something for the keyed version and just Integer for the non-keyed version.